src/pages/index.js

```diff
@@ -0,0 +1,330 @@
+  cardImageEl.alt = data.name;
+
+  if (data.isLiked) {
+    cardLikeButton.classList.add("card__like-button_liked");
+  }
+
+  cardLikeButton.addEventListener("click", (evt) => {
+    handleCardLike(evt, data._id);
+  });
+
+  cardDeleteButton.addEventListener("click", () => {
+    handleDeleteCard(cardElement, data._id);
+  });
+
+  cardImageEl.addEventListener("click", () => {
+    handleImageClick(data);
+  });
+
+  return cardElement;
+}
+
+function handleCardLike(evt, dataId) {
+  evt.target.classList.toggle("card__like-button_liked");
+
+  let isLiked = evt.target.classList.contains("card__like-button_liked")
+    ? true
+    : false;
+  api.toggleLike(dataId, isLiked).then().catch(console.error);
+}
+
+function handleImageClick(data) {
+  openModal(previewModal);
+  previewModalImageEl.src = data.link;
+  previewModalCaptionEl.textContent = data.name;
+  previewModalImageEl.alt = data.name;
+}
+
+function handleDeleteCard(cardElement, cardId) {
+  selectedCard = cardElement;
+  selectedCardId = cardId;
+  openModal(deleteModal);
+
+  deleteForm.addEventListener("submit", handleDeleteSubmit);
+}
+
+function handleDeleteSubmit(evt) {
+  evt.preventDefault();
+
+  const submitButton = evt.submitter;
+  submitButton.textContent = "Deleting...";
+
+  api
+    .deleteCard(selectedCardId)
+    .then(() => {
+      selectedCard.remove();
+    })
+    .catch(console.error)
+    .finally(() => {
+      submitButton.textContent = "Delete";
+      closeModal(deleteModal);
+    });
+}
+
+function handleOutsideClick(modal) {
+  return function (event) {
+    if (!event.target.closest(".modal__content")) {
+      closeModal(modal);
+    }
+  };
+}
+
+function handleEscapePress(modal) {
+  return function (event) {
+    if (event.key === "Escape" || event.key === "Esc") {
+      closeModal(modal);
+    }
+  };
+}
+
+let outsideClickHandler, escapePressHandler;
+
+function openModal(modal) {
+  modal.classList.add("modal_opened");
+
+  outsideClickHandler = handleOutsideClick(modal);
+  escapePressHandler = handleEscapePress(modal);
+
+  setTimeout(() => {
+    addListeners(outsideClickHandler, escapePressHandler);
+  }, 1);
+}
+
+function closeModal(modal) {
+  modal.classList.remove("modal_opened");
+
+  if (outsideClickHandler && escapePressHandler) {
+    removeListeners(outsideClickHandler, escapePressHandler);
+  }
+
+  outsideClickHandler = null;
+  escapePressHandler = null;
+}
+
+function addListeners(outsideClickHandler, escapePressHandler) {
+  document.addEventListener("click", outsideClickHandler);
+  document.addEventListener("keydown", escapePressHandler);
+}
+
+function removeListeners(outsideClickHandler, escapePressHandler) {
+  document.removeEventListener("click", outsideClickHandler);
+  document.removeEventListener("keydown", escapePressHandler);
+}
+
+function handleEditFormSubmit(evt) {
+  evt.preventDefault();
+
+  const submitButton = evt.submitter;
+  submitButton.textContent = "Saving...";
+  api
```

Comment:
**COULD BE IMPROVED**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **resetting** and **catching errors**


```js
// define a function for changing the button text. It accepts 4 params (the 2 last are optional with default texts)
export function renderLoading(isLoading, button, buttonText='Save', loadingText='Saving...') {
  if (isLoading) {
    button.textContent = loadingText
  } else {
    button.textContent = buttonText
  }
}

// define a universal function that accepts a request function, event and a default loading text 
function handleSubmit(request, evt, loadingText = 'Saving...') {
 // You need to prevent the default action in any submit handler
  evt.preventDefault();

  // the button is always available inside `event` as `submitter`
  const submitButton = evt.submitter;
  // fix the initial button text
  const initialText = submitButton.textContent;
  // change the button text before requesting
  renderLoading(true, submitButton, initialText, loadingText);
  // call the request function to be able to use the promise chain
  request()
    .then(() => {
      // any form should be reset after a successful response 
      // evt.target is the form in any submit handler		
      evt.target.reset();
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
    .catch(console.error)      

    // and in finally we need to stop loading
    .finally(() => {
      renderLoading(false, submitButton, initialText);
    });
}

```

Here is an example of handling the profile form submit:

```js
function handleProfileFormSubmit(evt) {
  // create a request function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally`
    return editProfile(nameInput.value, jobInput.value).then((userData) => {
      userName.textContent = userData.name;
      userJob.textContent = userData.about;
    });
  }
  // here we call handleSubmit passing the request and event (if you want a different loading text then you need to pass the 3rd param)
  handleSubmit(makeRequest, evt);
}

```

So, this way you can remove a lot of code duplicating. You will not need to search buttons, pass initial button texts and so on.

`handleSubmit` and `renderLoading` should be placed in `utils.js`, because they are utility functions

Reply:
It is interesting. I'm struggling to understand all this, so I don't want to break the code trying to figure out how to implement this.